js/eventloop/eventloop.go

```diff
@@ -191,12 +199,25 @@ func (e *EventLoop) Start(firstCallback func() error) error {
 }
 
 // WaitOnRegistered waits on all registered callbacks so we know nothing is still doing work.
+// This does call back the callbacks and more can be queued over time.
+// A different mechanism needs to be used to tell the users that the event loop has errored out or winding down for a
+// different reason.
 func (e *EventLoop) WaitOnRegistered() {
 	for {
-		_, awaiting := e.popAll()
-		if !awaiting {
-			return
+		queue, awaiting := e.popAll()
+		if len(queue) == 0 {
+			if !awaiting {
+				return
+			}
+			<-e.wakeupCh
+			continue
+		}
+
+		for _, f := range queue {
+			if err := f(); err != nil {
+				// TODO figure out if we should buffer all errors happening or send them on a channel
```

Comment:
I think a channel would be better here, so the caller can decide if wait or not for all of them. Maybe we could also stop to dequeue if the channel has been closed.

Reply:
The comment is about returning the errors to the caller.

If the caller *closes* the channel this will panic ;).

One of the reasons for this change is that there really isn't a reason to want to not run them, unless you want to kill the *whole* process. In which case the caller can just not call this or not wait for it to finish.

If the need ever arises for something additional we will add it.

This code definitely will need a lot more work around #2432 for example. But that issue has a lot of other changes that it will need.

Reply:
Did I understand you correctly @codebien ?

Reply:
>The comment is about returning the errors to the caller.

I got it, my proposal is around defining the API for listening about returned errors:
```
WaitOnRegistered(errs <-chan error) {
```

>If the caller closes the channel this will panic ;).

We can do a `select` on this eventual channel as usual.

>If the need ever arises for something additional we will add it.

Anyway, I mainly approved for this reason, at the moment, the demand is to have #2432 more than just an extension here. I was just commenting with my preference on the proposal in the TODO. We should not add this before a concrete vision on #2432.